### PR TITLE
compatible: Add arm64 support to build_charm.yaml and release_charm.yaml

### PR DESCRIPTION
**Pull request by carlcsaposs-canonical**
Uses spec (pending review) ST111 - Multi-architecture `upstream-source` in charm OCI resources
https://docs.google.com/document/d/19pzpza7zj7qswDRSHBlpqdBrA7Ndcnyh6_75cCxMKSo/edit

### CODE REVIEW

File: python/cli/data_platform_workflows_cli/collect_charm_bases.py [hunk 1]
```diff
@@ -21,7 +34,17 @@ def main():
...
     )
-    bases = [index for index, _ in enumerate(yaml_data["bases"])]
+    # GitHub runner for each base
+    runners = []
+    for base in yaml_data["bases"]:
+        # Bases format: https://discourse.charmhub.io/t/charmcraft-bases-provider-support/4713
+        architectures = (base.get("build-on") or base).get("architectures", ["amd64"])
+        assert (
+            len(architectures) == 1
+        ), f"Multiple architectures ({architectures}) in one (charmcraft.yaml) base not supported. Use one base per architecture"
```
Comment (taurus-forever) [inline python/cli/data_platform_workflows_cli/collect_charm_bases.py:44]: Does it mean charmcraft.yaml should have:
```
  - name: ubuntu
    channel: "22.04"
    architectures: [amd64]
  - name: ubuntu
    channel: "22.04"
    architectures: [arm64]
```

and not
```
  - name: ubuntu
    channel: "22.04"
    architectures: [amd64, arm64]
```

Comment (taurus-forever) [reply python/cli/data_platform_workflows_cli/collect_charm_bases.py:44]: Will we use `architectures: any` for workload-less charms? Or build/release two independent revisions there also?

Comment (dragomirp) [reply python/cli/data_platform_workflows_cli/collect_charm_bases.py:44]: > Will we use `architectures: any` for workload-less charms? Or build/release two independent revisions there also?

If there are binary dependencies (cryptography, psycopg, etc), we will need separate builds even if there is no workload.

Comment (taurus-forever) [reply python/cli/data_platform_workflows_cli/collect_charm_bases.py:44]: Sure, the question is about `s3-integrator` or `data-integrator`...
Will we produce more identical revision for different architectures based on each Git commit?

Comment (taurus-forever) [reply python/cli/data_platform_workflows_cli/collect_charm_bases.py:44]: It is exactly my question. Yes, we will always forget... newbies/rush/... 
We should create a charmhub improvement request, as today the WEB output us not useful:

Screenshot from 2024-02-26 14-48-24

Screenshot from 2024-02-26 14-47-54

Plus this visibility WEB confusion: https://github.com/canonical/charmhub.io/issues/1658


P.S. OK, the rule is: we should always clearly specify independent architectures even for workload-less charms.
Let's try to sell it @delgod on a sync call today. :crossed_fingers:

File: python/cli/data_platform_workflows_cli/charmcraft.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import enum
+
+
+class Architecture(str, enum.Enum):
+    X64 = "amd64"
```
Comment (dragomirp) [inline python/cli/data_platform_workflows_cli/charmcraft.py:5]: Might be better to name it `AMD64` to keep it consistent.

**Review by dragomirp — APPROVED** [review]

**Pull request merged**